scripts/data_overlap/ngram_hasher.py

```diff
@@ -0,0 +1,121 @@
+from typing import List, Tuple, Dict
+
+# A prime number chosen as a base for exponentiation, larger than 171,476 (estimated vocab size of English words)
+RABIN_KARP_PRIME = 172001
+
+# A prime number chosen as a base for hashing characters, larger than 256 (UTF-8 character space)
+CHAR_PRIME = 401
+
+hash_token_cache: Dict[str, int] = {}
+
+
+class RabinKarpHash:
+    """
+    Rabin-Karp rolling hash for hashing sequences of integers.
+
+    Initial hash:
+        For a List[int] i_0...i_n with a window w where w <= n,
+        the hash is calculated as:
+        i_0 * RABIN_KARP_PRIME^(w-1) + i_1 * RABIN_KARP_PRIME^(w-2) + ... + i_{w-1}
+
+    high_coefficient:
+        RABIN_KARP_PRIME^(w-1)
+    """
+
+    def __init__(self, int_sequence: List[int], window_size: int, mod: int = 8554560727166512717):
```

Comment:
Is mod any arbitrary large int?

Reply:
Yes, chosen to be large but small enough to fit in scala Long, and distinct from one another